src/linear_algebra/unitary_group.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Shing Tak Lam
+-/
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import data.complex.basic
+
+/-!
+# The Unitary Group
+
+This file defines elements of the unitary group `unitary_group n α`, where `α` is a `star_ring`.
+This consists of all `n` by `n` matrices with entries in `α` such that the star-transpose is its
+inverse. In addition, we define the group structure on `unitary_group n α`, and the embedding into
+the general linear group `general_linear_group α (n → α)`.
+
+We also define the orthogonal group `orthogonal_group n β`, where `β` is a `comm_ring`.
+
+## Main Definitions
+
+ * `matrix.unitary_group` is the type of matrices where the star-transpose is the inverse
+ * `matrix.unitary_group.group` is the group structure (under multiplication)
+ * `matrix.unitary_group.embedding_GL` is the embedding `unitary_group n α → GLₙ(α)`
+ * `matrix.orthogonal_group` is the type of matrices where the transpose is the inverse
+
+## References
+
+ * https://en.wikipedia.org/wiki/Unitary_group
+
+## Tags
+
+matrix group, group, unitary group, orthogonal group
+
+-/
+
+namespace matrix
+universes u v
+open linear_map
+open_locale matrix
+
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n]
+variables (α : Type v) [comm_ring α] [star_ring α]
```

Comment:
No need for this to be a `comm_ring` yet. `ring`, or even `semiring` just suffice for this file.